### ISSUE TITLE
boost can't bind udp socket on (maybe) tun interface with yggdrasil on it

### ISSUE THREAD

**Issue by r4sas** [PurpleI2P/i2pd#1275]
>Hi, I am having a problem that i2pd 2.21.1-1 stops working after a few reboots (or if day changes during poweroff?). Would this happen to be fixed on 2.22.0 which isn't in Debian repositories yet?

* https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=915501

_Originally posted by @Mikaela in https://github.com/PurpleI2P/i2pd/commit/779228857e5b212fbeff161dced60a61e1bc0c3e#commitcomment-31557786_

**Comment by r4sas** [comment]
I'll check as soon as possible in sid.

**Comment by l-n-s** [comment]
Not related, but when installing Ubuntu 18.04 ppa package there is a weird warning displayed:

```
Setting up libboost-program-options1.65.1:amd64 (1.65.1+dfsg-0ubuntu5) ...
Setting up libboost-filesystem1.65.1:amd64 (1.65.1+dfsg-0ubuntu5) ...
Setting up i2pd (2.22.0-1bionic1) ...
adduser: Warning: The home directory `/var/lib/i2pd' does not belong to the user you are currently creating.
Processing triggers for libc-bin (2.27-3ubuntu1) ...
Processing triggers for systemd (237-3ubuntu10.9) ...
```

**Comment by LLE8** [comment]
Please show 
`grep i2pd /etc/passwd`
`ls -ld /var/lib/i2pd`

**Comment by orignal** [comment]
LLE8 там еще openssl свинью подкладывает, пытаясь читать свой кофиг

**Comment by LLE8** [comment]
Да, у меня в Stretch такое же предупреждение. 
Воспроизвести просто
1 `apt-get purge i2pd`
2 `userdel -r i2pd`
3 `dpkg -i <path_to_you_package>/i2pd_2.22.0-1_amd64.deb`, если есть собственный пакет или  `apt-get install i2pd`, если из репозитория
Однако как мне кажется, там нет ничего страшного, т.к. потом выполняется usermod с нужными изменениями. можно выполнить `apt-get remove  i2pd` и `dpkg -i <path_to_you_package>/i2pd_2.22.0-1_amd64.deb` и увидеть `usermod: no changes`.

**Comment by r4sas** [comment]
Это нормально.
Но просьба не забивать иссуй посторонними сообщениями.

**Comment by r4sas** [comment]
Due to problems in debian-installer buster alpha4  I temporary can't check that issue.

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=916663

**Comment by Mikaela** [comment]
Are there any news? Debian Buster has been frozen so it shouldn't get major changes anymore and while nothing has been said in that bug report, I think it's possibly fixed regardless considering how the last comment says:

> This turns out to have been (expected) version skew between the alpha 4 
netboot image and the rest of the installer. It wasn't a bug.
Perhaps the documentation of how to use the netboot image
could be improved.

I am asking, because I saw email about new release.

**Comment by r4sas** [comment]
I still havent time to check it, maybe try again on this weekend.
To check that problem I need installed buter test-release.

**Comment by r4sas** [comment]
@Mikaela so, how to reproduce that issue? I'll test that on vbox, I need simply shutdown my virtual machine for day and try start it after time passed?

**Comment by Mikaela** [comment]
I think I installed I2PD in the evening, later powered the device off and in the morning it wasn't startable, but I am not entirely sure.

Do you have any ideas, @LLE8? I see that you opened a duplicate #1297 earlier.

**Comment by LLE8** [comment]
Can't reproduce that bug with 2.24  now in Debian Stretch.

**Comment by LLE8** [comment]
On another VDS 
i2pd 2.24, Debian Stretch
After 3 days running  `#shutdown -r now`
```
root@vds:~# systemctl status i2pd.service
  i2pd.service - I2P Router written in C++
   Loaded: loaded (/lib/systemd/system/i2pd.service; disabled; vendor preset: enabled)
   Active: inactive (dead)
     Docs: man:i2pd(1)
           https://i2pd.readthedocs.io/en/latest/
```
That is it: "service; disabled; vendor preset: enabled"?
Nothing interesting in logs. What to look for?

**Comment by r4sas** [comment]
Currently I can't reproduce this on clean buster install with i2pd 2.23.0 package from official repo.

```
● i2pd.service - I2P Router written in C++
   Loaded: loaded (/lib/systemd/system/i2pd.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2019-03-26 18:56:27 UTC; 2min 23s ago
     Docs: man:i2pd(1)
           https://i2pd.readthedocs.io/en/latest/
 Main PID: 636 (i2pd)
    Tasks: 15 (limit: 1150)
   Memory: 18.1M
   CGroup: /system.slice/i2pd.service
           └─636 /usr/sbin/i2pd --conf=/etc/i2pd/i2pd.conf --tunconf=/etc/i2pd/tunnels.conf --tunnelsdir=/etc/i2pd/
```

**Comment by Mikaela** [comment]
I had the service disabled when I previously commented and I enabled and started it and at that time it worked. Later I powered off the laptop while taking a shower and doing other things and now I remembered this again and

```
● i2pd.service - I2P Router written in C++
   Loaded: loaded (/lib/systemd/system/i2pd.service; enabled; vendor preset: disabled)
   Active: failed (Result: signal) since Tue 2019-03-26 21:32:41 EET; 1h 6min ago
     Docs: man:i2pd(1)
           https://i2pd.readthedocs.io/en/latest/
  Process: 864 ExecStart=/usr/sbin/i2pd --conf=/etc/i2pd/i2pd.conf --tunconf=/etc/i2pd/tunnels.conf --tunnelsdir=/etc/i2pd/tunnels.conf.d --pidfile=/var/run/i
 Main PID: 952 (code=killed, signal=ABRT)

Mar 26 21:32:09 sedric systemd[1]: Starting I2P Router written in C++...
Mar 26 21:32:09 sedric systemd[864]: RuntimeDirectory 'i2pd' already exists but the mode is different. (filesystem: 755 RuntimeDirectoryMode: 700)
Mar 26 21:32:09 sedric systemd[864]: LogsDirectory 'i2pd' already exists but the mode is different. (filesystem: 755 LogsDirectoryMode: 700)
Mar 26 21:32:13 sedric systemd[1]: Started I2P Router written in C++.
Mar 26 21:32:41 sedric systemd[1]: i2pd.service: Main process exited, code=killed, status=6/ABRT
Mar 26 21:32:41 sedric systemd[1]: i2pd.service: Failed with result 'signal'.
```

```
i2pd:
  Installed: 2.23.0-1
  Candidate: 2.23.0-1
  Version table:
 *** 2.23.0-1 990
        990 https://deb.debian.org/debian testing/main amd64 Packages
        -10 https://deb.debian.org/debian sid/main amd64 Packages
        100 /var/lib/dpkg/status
```

```
Debian GNU/Linux buster/sid
```

**Comment by r4sas** [comment]
Can you try temporary disable apparmor profile `usr.sbin.i2pd`?

**Comment by Mikaela** [comment]
No change.

```
┌┌(root@sedric)┌(2872/pts/1)┌(11:53:03/26/19)┌-                                                                                                               
└┌(#:~)┌- aa-disable i2pd                                                                                                                                     
Disabling /usr/sbin/i2pd.                                                    
┌┌(root@sedric)┌(2873/pts/1)┌(11:54:03/26/19)┌-
└┌(#:~)┌- systemctl start i2pd
┌┌(root@sedric)┌(2874/pts/1)┌(11:54:03/26/19)┌-
└┌(#:~)┌- systemctl status i2pd
● i2pd.service - I2P Router written in C++
   Loaded: loaded (/lib/systemd/system/i2pd.service; enabled; vendor preset: disabled)
   Active: failed (Result: signal) since Tue 2019-03-26 23:54:07 EET; 2s ago
     Docs: man:i2pd(1)
           https://i2pd.readthedocs.io/en/latest/
  Process: 3703 ExecStart=/usr/sbin/i2pd --conf=/etc/i2pd/i2pd.conf --tunconf=/etc/i2pd/tunnels.conf --tunnelsdir=/etc/i2pd/tunnels.conf.d --pidfile=/var/run$
 Main PID: 3704 (code=killed, signal=ABRT)
 
Mar 26 23:54:07 sedric systemd[1]: Starting I2P Router written in C++...
Mar 26 23:54:07 sedric systemd[3703]: LogsDirectory 'i2pd' already exists but the mode is different. (filesystem: 755 LogsDirectoryMode: 700)
Mar 26 23:54:07 sedric systemd[1]: Started I2P Router written in C++.
Mar 26 23:54:07 sedric systemd[1]: i2pd.service: Main process exited, code=killed, status=6/ABRT
Mar 26 23:54:07 sedric systemd[1]: i2pd.service: Failed with result 'signal'.
...skipping...
● i2pd.service - I2P Router written in C++
   Loaded: loaded (/lib/systemd/system/i2pd.service; enabled; vendor preset: disabled)
   Active: failed (Result: signal) since Tue 2019-03-26 23:54:07 EET; 2s ago
     Docs: man:i2pd(1)                                                                
           https://i2pd.readthedocs.io/en/latest/                           
  Process: 3703 ExecStart=/usr/sbin/i2pd --conf=/etc/i2pd/i2pd.conf --tunconf=/etc/i2pd/tunnels.conf --tunnelsdir=/etc/i2pd/tunnels.conf.d --pidfile=/var/run$
 Main PID: 3704 (code=killed, signal=ABRT)

Mar 26 23:54:07 sedric systemd[1]: Starting I2P Router written in C++...
Mar 26 23:54:07 sedric systemd[3703]: LogsDirectory 'i2pd' already exists but the mode is different. (filesystem: 755 LogsDirectoryMode: 700)
Mar 26 23:54:07 sedric systemd[1]: Started I2P Router written in C++.
Mar 26 23:54:07 sedric systemd[1]: i2pd.service: Main process exited, code=killed, status=6/ABRT
Mar 26 23:54:07 sedric systemd[1]: i2pd.service: Failed with result 'signal'.
```

(Good night, it's 23:55 here and I have early wake-up)

**Comment by r4sas** [comment]
`netstat -aonp | grep -P "4444|4447|7070|7656"`
and 
`tail -n 15 /var/log/i2pd/i2pd.log`
(please remove sensitive info before posting if any exists).

**Comment by Mikaela** [comment]
> `netstat -aonp | grep -P "4444|4447|7070|7656"`

no output

>  `tail -n 15 /var/log/i2pd/i2pd.log`

```
07:34:06@318/info - i2pd v2.23.0 starting
07:34:06@318/debug - FS: main config file: /etc/i2pd/i2pd.conf
07:34:06@318/debug - FS: data directory: /var/lib/i2pd
07:34:06@318/info - AESNI enabled
07:34:06@318/info - AVX enabled
07:34:06@318/info - Daemon: accepting incoming connections at port 2743
07:34:06@318/info - Daemon: bandwidth set to 256KBps
07:34:06@318/info - Daemon: using system limit in 4096 max open files
07:34:06@318/info - Daemon: starting NetDB
07:34:07@318/info - Family: 5 certificates loaded
07:34:27@318/error - Family: 347f69139bc02660c00fb9694c7561a382d6a637708638f3480306482d233ba0 is too long
07:34:27@318/warn - RouterInfo: family signature verification failed
07:34:31@318/info - NetDb: 1036 routers loaded (545 floodfils)
07:34:31@318/info - Daemon: starting Transports
07:34:31@318/info - NTCP: Start listening V6 TCP port 2743
```

I don't have any config in i2pd that I know of so I can start from the beginning if something happened to be sensitive.

**Comment by r4sas** [comment]
What happen if you start `i2pd` from your user account?
You can copy `i2pd.conf ` and `certificates` folder in `~/.i2pd/` before starting.

**Comment by Mikaela** [comment]
When I start it as my account, it seems to work.

Where should coredump be generated?

**Comment by r4sas** [comment]
`/var/lib/i2pd` file named `core`

**Comment by Mikaela** [comment]
I am not sure if it worked as now I get no output for some reaosn. For the previous comment I apparently copied over pidfile(?) so it probably ran without config, it's midnight here and I am half-asleep.

I find two core files, 

```gdb
No stack.
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
        set = {__val = {0, 18446744073709551615 <repeats 12 times>, 0, 0, 0}}
        pid = <optimized out>
        tid = <optimized out>
        ret = <optimized out>
#1  0x00007ff9dbd855fe in __GI_abort () at abort.c:100
        act = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, sa_mask = {__val = {18446744073709551615 <repeats 16 times>}}, sa_flags = 0, 
          sa_restorer = 0x0}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00007ff9dbfeb983 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#3  0x00007ff9dbff18e6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#4  0x00007ff9dbff1921 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#5  0x00007ff9dbff1b54 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#6  0x000055670d7d1e35 in void boost::throw_exception<boost::system::system_error>(boost::system::system_error const&) ()
No symbol table info available.
#7  0x000055670d7d1ef2 in boost::asio::detail::do_throw_error(boost::system::error_code const&, char const*) ()
No symbol table info available.
#8  0x000055670d906308 in i2p::transport::SSUServer::OpenSocketV6() ()
No symbol table info available.
#9  0x000055670d909763 in i2p::transport::SSUServer::SSUServer(boost::asio::ip::address const&, int) ()
No symbol table info available.
#10 0x000055670d8a04a6 in i2p::transport::Transports::Start(bool, bool) ()
No symbol table info available.
#11 0x000055670d7b337d in i2p::util::Daemon_Singleton::start() ()
No symbol table info available.
#12 0x000055670d802e73 in i2p::util::DaemonLinux::start() ()
No symbol table info available.
#13 0x000055670d7a7cfb in main ()
No symbol table info available.
```

```gdb
No stack.                                                                                                                                            
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
        set = {__val = {0, 18446744073709551615 <repeats 12 times>, 0, 0, 0}}                                                                                
        pid = <optimized out>   
        tid = <optimized out>                                                                                                                                
        ret = <optimized out>   
#1  0x00007ff9dbd855fe in __GI_abort () at abort.c:100                                                                                                       
        act = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, sa_mask = {__val = {18446744073709551615 <repeats 16 times>}}, sa_flags = 0,
          sa_restorer = 0x0}                                  
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00007ff9dbfeb983 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.    
#3  0x00007ff9dbff18e6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6                                                                                
No symbol table info available. 
#4  0x00007ff9dbff1921 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6                                                                  
No symbol table info available.
#5  0x00007ff9dbff1b54 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#6  0x000055670d7d1e35 in void boost::throw_exception<boost::system::system_error>(boost::system::system_error const&) ()
No symbol table info available. 
#7  0x000055670d7d1ef2 in boost::asio::detail::do_throw_error(boost::system::error_code const&, char const*) ()
No symbol table info available.    
#8  0x000055670d906308 in i2p::transport::SSUServer::OpenSocketV6() ()
No symbol table info available. 
#9  0x000055670d909763 in i2p::transport::SSUServer::SSUServer(boost::asio::ip::address const&, int) ()
No symbol table info available.
#10 0x000055670d8a04a6 in i2p::transport::Transports::Start(bool, bool) ()
No symbol table info available.
#11 0x000055670d7b337d in i2p::util::Daemon_Singleton::start() ()
No symbol table info available.
#12 0x000055670d802e73 in i2p::util::DaemonLinux::start() ()
No symbol table info available.
#13 0x000055670d7a7cfb in main ()
No symbol table info available.
#0  __libc_read (nbytes=8191, buf=0x55db27f1bcc0, fd=5) at ../sysdeps/unix/sysv/linux/read.c:26
        resultvar = 808
        sc_cancel_oldtype = 0
        sc_ret = <optimized out>
        sc_ret = <optimized out>
        sc_ret = <optimized out>
        resultvar = <optimized out>
        resultvar = <optimized out>                                                                                                                           
        __arg3 = <optimized out>
        __arg2 = <optimized out>
        __arg1 = <optimized out>
        _a3 = <optimized out>
        _a2 = <optimized out>
        _a1 = <optimized out>
        sc_cancel_oldtype = <optimized out>
        resultvar = <optimized out>
        resultvar = <optimized out>
        __arg3 = <optimized out>
        __arg2 = <optimized out>
        __arg1 = <optimized out>
        _a3 = <optimized out>
        _a2 = <optimized out>
        _a1 = <optimized out>
#1  __libc_read (fd=5, buf=0x55db27f1bcc0, nbytes=8191) at ../sysdeps/unix/sysv/linux/read.c:24
        sc_ret = <optimized out>
        resultvar = <optimized out>
        resultvar = <optimized out>
        __arg3 = <optimized out>
        __arg2 = <optimized out>
        __arg1 = <optimized out>
        _a3 = <optimized out>
        _a2 = <optimized out>
        _a1 = <optimized out>
        sc_cancel_oldtype = <optimized out>
        resultvar = <optimized out>
        resultvar = <optimized out>
        __arg3 = <optimized out>
        __arg2 = <optimized out>
        __arg1 = <optimized out>
        _a3 = <optimized out>
        _a2 = <optimized out>
        _a1 = <optimized out>
#2  0x00007f8cd6896fb7 in std::__basic_file<char>::xsgetn(char*, long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#3  0x00007f8cd68ce420 in std::basic_filebuf<char, std::char_traits<char> >::underflow() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#4  0x00007f8cd6900c96 in std::basic_streambuf<char, std::char_traits<char> >::xsgetn(char*, long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007f8cd68cec23 in std::basic_filebuf<char, std::char_traits<char> >::xsgetn(char*, long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#6  0x00007f8cd68db98d in std::istream::read(char*, long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#7  0x000055db27b6eef4 in i2p::data::RouterInfo::LoadFile() ()
No symbol table info available.
#8  0x000055db27b752a9 in i2p::data::RouterInfo::ReadFromFile() ()
No symbol table info available.
#9  0x000055db27b753ef in i2p::data::RouterInfo::RouterInfo(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
No symbol table info available.
#10 0x000055db27b48b2c in i2p::data::NetDb::LoadRouterInfo(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
No symbol table info available.
#11 0x000055db27b4903b in i2p::data::NetDb::Load() ()
No symbol table info available.
#12 0x000055db27b4915e in i2p::data::NetDb::Start() ()
No symbol table info available.
#13 0x000055db27ab1261 in i2p::util::Daemon_Singleton::start() ()
No symbol table info available.
#14 0x000055db27b00e73 in i2p::util::DaemonLinux::start() ()
No symbol table info available.
#15 0x000055db27aa5cfb in main ()
No symbol table info available.
```

I cannot find a `i2pd-dbgsym` package, should one exist or which packages do I need? I think https://wiki.debian.org/AutomaticDebugPackages is the documentation for packagers.

Good night

**Comment by r4sas** [comment]
> ```gdb
> #7  0x000055670d7d1ef2 in boost::asio::detail::do_throw_error(boost::system::error_code const&, char const*) ()
> No symbol table info available.    
> #8  0x000055670d906308 in i2p::transport::SSUServer::OpenSocketV6() ()
> ```

That's make sense.

> I cannot find a i2pd-dbgsym package, should one exist or which packages do I need?

About debug package better ask @yangfl, I really don't know why he removed it from building.

**Comment by LLE8** [comment]
@Mikaela 
First. 
You may build your own i2pd deb package, with you own custom i2pd.conf and custom i2pd.service inside package, it is not difficult. https://i2pd.readthedocs.io/en/latest/devs/building/unix/ .
In your local source tree uncomment `LimitCORE=infinity` in contrib/i2pd.service to enable coredumps out of box. Also you may change some lines in contrib/i2pd.conf.  After that run `debuild --no-tgz-check -us -uc -b`. You will need about 2 GB RAM or swap. 
As result you will get two packages: program `i2pd_2.24-***.deb` and  symbols `i2pd-dbg_2.24-***.deb`. Also you may change names of core dump files, tune /proc/sys/kernel/core_pattern  http://man7.org/linux/man-pages/man5/core.5.html .

**Comment by r4sas** [comment]
@LLE8 she already got coredump, all is needed - install dbgsym package from debian-debug repository and get new stack trace.

**Comment by LLE8** [comment]
@r4sas   I think it's false target.

**Comment by r4sas** [comment]
@Mikaela can you provide full backtrace with dbgsym packages installed?

**Comment by Mikaela** [comment]
I am having bit of a situation with mental health, long todo lists and deadlines, I will try to take a look at this when I can, possibly on Monday at earliest.

**Comment by r4sas** [comment]
Ok, no rush.

**Comment by Mikaela** [comment]
```gdb
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
        set = {__val = {0, 18446744073709551615 <repeats 12 times>, 0, 38, 
            1}}
        pid = <optimized out>
        tid = <optimized out>
        ret = <optimized out>
#1  0x00007f825cfea5fe in __GI_abort () at abort.c:100
        act = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, 
          sa_mask = {__val = {18446744073709551615 <repeats 16 times>}}, 
          sa_flags = 0, sa_restorer = 0x0}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00007f825d3b4983 in __gnu_cxx::__verbose_terminate_handler ()
    at ../../../../src/libstdc++-v3/libsupc++/vterminate.cc:95
        terminating = true
        t = <optimized out>
#3  0x00007f825d3ba8c6 in __cxxabiv1::__terminate (handler=<optimized out>)
    at ../../../../src/libstdc++-v3/libsupc++/eh_terminate.cc:47
No locals.
#4  0x00007f825d3ba901 in std::terminate ()
    at ../../../../src/libstdc++-v3/libsupc++/eh_terminate.cc:57
No locals.
#5  0x00007f825d3bab34 in __cxxabiv1::__cxa_throw (
    obj=obj@entry=0x55ab87c2f3a0, 
    tinfo=tinfo@entry=0x55ab8745f9c0 <typeinfo for boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >>, 
    dest=dest@entry=0x55ab8722efb0 <boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >::~clone_impl()>) at ../../../../src/libstdc++-v3/libsupc++/eh_throw.cc:95
        globals = <optimized out>
        header = 0x55ab87c2f320
#6  0x000055ab872345bd in boost::throw_exception<boost::system::system_error>
    (e=...) at /usr/include/boost/exception/exception.hpp:77
No locals.
#7  0x000055ab87234672 in boost::asio::detail::do_throw_error (err=..., 
    location=location@entry=0x55ab873ce72b "bind")
    at /usr/include/boost/asio/detail/impl/throw_error.ipp:38
        e = {<std::runtime_error> = {<std::exception> = {
              _vptr.exception = 0x55ab8745f168 <vtable for boost::system::system_error+16>}, _M_msg = "bind"}, m_error_code = {m_val = 99, 
            m_cat = 0x7f825d58e070}, m_what = ""}
#8  0x000055ab873155a5 in boost::asio::detail::throw_error (
    location=0x55ab873ce72b "bind", err=...)
    at /usr/include/boost/asio/detail/throw_error.hpp:42
No locals.
#9  boost::asio::basic_socket<boost::asio::ip::udp>::bind (endpoint=..., 
    this=0x55ab87bfccd0) at /usr/include/boost/asio/basic_socket.hpp:699
        ec = {m_val = 99, m_cat = 0x7f825d58e070}
        ec = <optimized out>
#10 i2p::transport::SSUServer::OpenSocketV6 (this=0x55ab87bfcbf0)
    at libi2pd/SSU.cpp:60
No locals.
#11 0x000055ab87316944 in i2p::transport::SSUServer::SSUServer (
    this=0x55ab87bfcbf0, addr=..., port=<optimized out>)
    at libi2pd/SSU.cpp:24
No locals.
#12 0x000055ab872974b6 in i2p::transport::Transports::Start (
    this=this@entry=0x55ab87468a80 <i2p::transport::transports>, 
    enableNTCP=true, enableSSU=true) at libi2pd/Transports.cpp:233
        address = 
    std::shared_ptr<i2p::data::RouterInfo::Address> (use count 1, weak count 0) = {get() = 0x55ab87beb1d0}
        __for_range = std::__cxx11::list = {
          [0] = std::shared_ptr<i2p::data::RouterInfo::Address> (use count 1, weak count 0) = {get() = 0x55ab87bfd470}, 
          [1] = std::shared_ptr<i2p::data::RouterInfo::Address> (use count 1, weak count 0) = {get() = 0x55ab87beb1d0}}
        __for_begin = <optimized out>
        __for_end = <optimized out>
        ntcpproxy = ""
        proxyurl = {schema = "", user = "", pass = "", host = "", port = 0, 
          path = "", query = "", frag = ""}
        softLimit = 0
        hardLimit = 0
        threads = 1
        ntcp2 = true
        addresses = std::__cxx11::list = {
          [0] = std::shared_ptr<i2p::data::RouterInfo::Address> (use count 1, weak count 0) = {get() = 0x55ab87bfd470}, 
          [1] = std::shared_ptr<i2p::data::RouterInfo::Address> (use count 1, weak count 0) = {get() = 0x55ab87beb1d0}}
#13 0x000055ab8721393d in i2p::util::Daemon_Singleton::start (
    this=this@entry=0x55ab87468360 <i2p::util::DaemonLinux::Instance()::instance>) at daemon/Daemon.cpp:300
        upnp = false
        nettime = false
        ntcp = true
        ssu = true
        http = 105
        i2pcontrol = 50
#14 0x000055ab87250669 in i2p::util::DaemonLinux::start (
    this=0x55ab87468360 <i2p::util::DaemonLinux::Instance()::instance>)
    at daemon/UnixDaemon.cpp:172
        limit = {rlim_cur = 4096, rlim_max = 4096}
        nfiles = 0
        cfsize = 0
        pidfile = "/var/run/i2pd/i2pd.pid"
        sa = {__sigaction_handler = {
            sa_handler = 0x55ab8724fba0 <handle_signal(int)>, 
            sa_sigaction = 0x55ab8724fba0 <handle_signal(int)>}, sa_mask = {
            __val = {0 <repeats 16 times>}}, sa_flags = 268435456, 
          sa_restorer = 0x55ab87bf9cd0}
#15 0x000055ab8720a31b in main (argc=8, argv=0x7fffd21c8138)
    at daemon/i2pd.cpp:23
No locals.
```

**Comment by Mikaela** [comment]
Did the traceback say anything useful or is there anything I can do to get useful information?

**Comment by r4sas** [comment]
It said only that crash happens when boost tried to bind to ipv6 address...
```
#9  boost::asio::basic_socket<boost::asio::ip::udp>::bind (endpoint=..., 
    this=0x55ab87bfccd0) at /usr/include/boost/asio/basic_socket.hpp:699
        ec = {m_val = 99, m_cat = 0x7f825d58e070}
        ec = <optimized out>```

**Comment by Mikaela** [comment]
I don't have a public IPv4 address, I am behind CGN. Wouldn't that prevent me from using I2Pd?

I am using [Yggdrasil network](https://yggdrasil-network.github.io/), is that potentially causing it? It appears a a IPv6 address in ifconfig (in addition to my clearnet address), e.g.:

```
yggdrasil: flags=67<UP,BROADCAST,RUNNING>  mtu 65521
        inet6 fe80::...  prefixlen 64  scopeid 0x20<link>
        inet6 201:b060:d3fe:dbef:8f6d:561d:6892:9cd4  prefixlen 7  scopeid 0x0<global>
        ether xx:...:xx  txqueuelen 1000  (Ethernet)
        RX packets 2352  bytes 1181665 (1.1 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 5221  bytes 546693 (533.8 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

**Comment by r4sas** [comment]
It can be... I really dont know how boost handle such networks (which is not in 2000::/3 range).

**Comment by r4sas** [comment]
Can you try set `ifname6` to your network interface with real IPv6 address?
btw, if router.info exists, can you see which IPv6 address it contain, from real network or ygg?

**Comment by Mikaela** [comment]
> Can you try set ifname6 to your network interface name with real IPv6 address?

Yes, i2pd seems to start without a problem and I see the correct address in "Our external address:"

>  btw, if router.info exists, can you see which IPv6 address it contain, from real network or ygg?

Where should it be located? I should probably shut down I2Pd and start it with the crashing settings?

Should the issue be renamed now that it's known that it's not related to reboot/day change?

**Comment by r4sas** [comment]
> Where should it be located? I should probably shut down I2Pd and start it with the crashing settings?

Now its not necessary, because it was overwritten with new address.

> Should the issue be renamed now that it's known that it's not related to reboot/day change?

Ok, lets rename it, as now reason is known. Maybe we will add filters to avoid such bindings.

**Comment by Mikaela** [comment]
It looks like this happened again this morning. My router.info shows `2001:999:42:19db:ac6d:b9d8:9c75:bdd9`, maybe I2Pd wanted to use it before WLAN connected?

`-rw-r--r-- 1 i2pd i2pd 766 May  8 10:38 router.info`

Oh and my IPv6 actually seems to be `2001:999:22:1c9:a030:b686:b111:60e7`, but it's possible privacy extensions changed it since boot.

**Comment by r4sas** [comment]
Dropping router.info helps?

**Comment by LLE8** [comment]
Мне показалось, что с yggdrasil вообще все как-то странно
https://github.com/PurpleI2P/i2pd/issues/837#issuecomment-487321058

**Comment by r4sas** [comment]
Как видишь, boost не может почему-то цепляться к подобным интерфейсам с нестандартными адресными пространствами./
Надо будет либо отбрасывать подобные интерфейсы, либо изучать более глубу причину на стороне буста.

**Comment by Mikaela** [comment]
> Dropping router.info before starting helps?

If you mean moving `/var/lib/i2pd/router.info`, that and `systemctl start i2pd` lets i2pd start again.

**Comment by LLE8** [comment]
Значит и поверх cjdns тоже нормально работать не будет?
И еще через 19 суток работы заметил, что один из потоков i2pd жрет cpu на 97% при трафике 10-20-30 Кбит/с

**Comment by orignal** [comment]
Поток не NTCP случайно?

**Comment by LLE8** [comment]
Не скажу, уже убил, машинка понадобилась для новых экспериментов.
А как различать потоки выполняющегося процесса по на назначению? Который из них NTCP, а который SSU?

**Comment by orignal** [comment]
цепляться отладчиком

**Comment by LLE8** [comment]
Снова вижу более 95% cpu один поток. Открыть новый баг?

**Comment by orignal** [comment]
нет. отключить NTCP и посмотреть. это известная проблема с ним

**Comment by LLE8** [comment]
Да, похоже, отключение ntcp помогло

**Comment by orignal** [comment]
В следующем релизе планируем его отключить по умолчанию, а потом и убрать совсем.

**Comment by Mikaela** [comment]
Have there been any news related to this issue? Could it be renamed to "Bind failure to IPv6 address prevents I2Pd from starting when IPv6 address changes"?

**Comment by r4sas** [comment]
All testing stuff currently in #1349, but currently I havent time to finish it. Now I'm waiting for NTCP removal before continuation.

**Comment by r4sas** [comment]
Needs testing with current codebase, ntcp was removed in 2.34.0.

**Comment by r4sas** [comment]
Yggdrasil support will be added in 2.36.0.
Current codebase can react different way than before, so I closing this issue. If any new problems with binding of UDP, please open new issue.
